pkg/apis/core/v1alpha1/types_federatedobject.go

```diff
@@ -0,0 +1,254 @@
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired behavior of the ClusterFederatedObject.
+	Spec GenericFederatedObjectSpec `json:"spec"`
+
+	// Status describes the most recently observed status of the ClusterFederatedObject.
+	// +optional
+	Status GenericFederatedObjectStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterFederatedObjectList contains a list of ClusterFederatedObject.
+type ClusterFederatedObjectList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ClusterFederatedObject `json:"items"`
+}
+
+// GenericFederatedObjectSpec defines the desired behavior of a FederatedObject or ClusterFederatedObject.
+type GenericFederatedObjectSpec struct {
+	// Template is the base template of the Kubernetes object to be propagated.
+	Template apiextensionsv1.JSON `json:"template"`
+	// Overrides describe the overrides that should be applied to the base template of the Kubernetes object before it
+	// is propagated to individual member clusters.
+	Overrides []OverrideWithController `json:"overrides"`
+	// Placements describe the member clusters that the Kubernetes object will be propagated to, which is a union of all
+	// the listed clusters.
+	Placements []PlacementWithController `json:"placements"`
+	// Follows defines other objects, or "leaders", that the Kubernetes object should follow during propagation. I.e.
+	// the Kubernetes object should be propagated to all member clusters that its "leaders" are placed in.
+	Follows []LeaderReference `json:"follows"`
+}
+
+// GenericFederatedObjectStatus describes the most recently observed status of a FederatedObject or ClusterFederatedObject.
+type GenericFederatedObjectStatus struct {
+	// SyncedGeneration is the generation of this FederatedObject when it was last synced to selected member clusters.
+	SyncedGeneration int64 `json:"syncedGeneration,omitempty"`
+	// Conditions describe the current state of this FederatedObject.
+	Conditions []GenericFederatedObjectCondition `json:"conditions,omitempty"`
+	// Clusters contains the propagation status of the Kubernetes object for individual member clusters.
+	Clusters []PropagationStatus `json:"clusters,omitempty"`
+
+	// CollisionCount can be used in conjunction with RevisionHistory to implement rollbacks.
+	// +optional
+	CollisionCount *int32 `json:"collisionCount,omitempty"`
+}
+
+// PlacementWithController describes the member clusters that a Kubernetes object should be propagated to.
+type PlacementWithController struct {
+	// Controller identifies the controller responsible for this placement.
+	Controller string `json:"controller"`
+	// Placement is the list of member clusters that the Kubernetes object should be propagated to.
+	Placement []ClusterReference `json:"placement"`
+}
+
+// ClusterReference represents a single member cluster.
+type ClusterReference struct {
```

Comment:
Should we flatten this struct now that we don't support clusterselector anymore?

Reply:
i think keeping it this way might be good too, makes it evident and self-explanatory what a `Placement` is, rather than a list of strings